idunn/datasources/pages_jaunes.py

```diff
@@ -58,21 +77,78 @@ def get_places_bbox(self, raw_categories, bbox, size=10, query=""):
-
+    def get_place(self, poi_id):
         es_places = self.es.search(
             index=self.es_index,
-            body={"query": {"bool": {"filter": {"term": {"_id": internal_id}}}}},
+            body={"query": {"bool": {"filter": {"term": {"_id": self.internal_id(poi_id)}}}}},
             ignore_unavailable=True,
         )
 
         es_place = es_places.get("hits", {}).get("hits", [])
         if len(es_place) == 0:
-            raise HTTPException(status_code=404, detail=f"place {id} not found")
+            raise HTTPException(status_code=404, detail=f"place {poi_id} not found")
         if len(es_place) > 1:
-            logger.warning("Got multiple places with id %s", id)
+            logger.warning("Got multiple places with id %s", poi_id)
         return PjPOI(es_place[0]["_source"])
 
 
-pj_source = PjSource()
+class ApiPjSource(PjSource):
+    PJ_INFO_API_URL = "https://api.pagesjaunes.fr/v1/pros"
+    PJ_FIND_API_URL = "https://api.pagesjaunes.fr/v1/pros/search"
+    PJ_AUTHORIZATION = "https://api.pagesjaunes.fr/oauth/client_credential/accesstoken"
+
+    def __init__(self):
+        self.enabled = True
+        self.client_id = settings.get("PJ_API_ID")
+        self.client_secret = settings.get("PJ_API_SECRET")
+
+        # TODO: refresh token if available?
+        #       https://requests-oauthlib.readthedocs.io/en/latest/oauth2_workflow.html#refreshing-tokens
+        self.client = BackendApplicationClient(client_id=self.client_id)
+        self.oauth = OAuth2Session(client=self.client)
+        self.token = self.oauth.fetch_token(
+            token_url=self.PJ_AUTHORIZATION,
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+        )
+
+    @staticmethod
+    def format_where(bbox):
+        left, bot, right, top = bbox
+        return f"gZ{left},{top},{right},{bot}"
+
+    # TODO: this requires a strong error management as it calls an external API
+    def get_from_params(self, url, params) -> PjApiPOI:
+        res = self.oauth.get(url, params=params)
+        res.raise_for_status()
+        return res.json()
+
+    def get_places_bbox(self, raw_categories, bbox, size=10, query="") -> List[PjApiPOI]:
+        query_params = {
+            "what": raw_categories,
+            "where": self.format_where(bbox),
+            # TODO: add some scrolling mechanism, which should come with some async?
+            "max": min(30, size),
+        }
+
+        if query:
+            query_params["q"] = query
+
+        res = pj_find.Response(**self.get_from_params(self.PJ_FIND_API_URL, query_params))
+
+        if not res:
+            return None
+
+        return [PjApiPOI(listing) for listing in res.search_results.listings or []]
+
+    def get_place(self, poi_id) -> PjApiPOI:
+        return PjApiPOI(
+            pj_info.Response(
+                **self.get_from_params(
+                    self.PJ_INFO_API_URL, {"listing_id": self.internal_id(poi_id)}
```

Comment:
Same comment: `poi_id` is based on the `merchant_id`, that needs to be passed in the URL path (and not as a query param).

Reply:
Oh right, then the business info API has to be called using the positional `pro_id` parameter (which appears to be equivalent to `merchant_id`? I'll add a comment in my list), that's what go me confused as `?merchant_id=XXXXXXXX` leads to empty results.